File: mock-csharp/MockRuntime.Core/Engine.cs
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace MockRuntime;

public sealed partial class RuntimeEngine(IMockDebugger? debugger = null)
{
    string sourceFile = string.Empty;
    string[] sourceLines = Array.Empty<string>();
    int currentLine; // zero-based
    int? currentColumn; // 0-based when set

    readonly Dictionary<string, List<Breakpoint>> bps = new(StringComparer.OrdinalIgnoreCase);
    int nextBpId = 1;

    // Instruction stream and per-line ranges
    readonly List<Word> instructions = new();
    readonly List<int> starts = new();
    readonly List<int> ends = new();
    int instruction;
    volatile bool paused;

    // Data/instruction breakpoints and exception filters
    readonly HashSet<int> instructionBreakpoints = new();
    readonly Dictionary<string, string> breakAddresses = new(StringComparer.OrdinalIgnoreCase);
    string? namedException;
    bool otherExceptions;

    // Simple variable tracking for read/write detection
    readonly HashSet<string> variables = new(StringComparer.OrdinalIgnoreCase);
    readonly Dictionary<string, object?> locals = new(StringComparer.OrdinalIgnoreCase);
    IMockDebugger? debugger = debugger;
    
    public IMockDebugger? Debugger
    {
        get => debugger;
        set => debugger = value;
    }
    
    readonly record struct Breakpoint(int Id, int Line, bool Verified);

    public string SourceFile => sourceFile;
    public int SourceLength => sourceLines.Length;

    public async Task LoadSourceAsync(string path)
    {
        sourceFile = NormalizePath(path);
        sourceLines = await File.ReadAllLinesAsync(sourceFile, Encoding.UTF8);
        currentLine = 0;
        currentColumn = null;

        // rebuild instruction map
        instructions.Clear();
        starts.Clear();
        ends.Clear();
        for (int l = 0; l < sourceLines.Length; l++)
        {
            starts.Add(instructions.Count);
            foreach (var w in GetWords(l, sourceLines[l]))
            {
                instructions.Add(w);
            }
            ends.Add(instructions.Count);
        }
        instruction = starts.Count > 0 ? starts[0] : 0;

        // Now that a source is loaded, re-verify any pending breakpoints for this file
        ReverifyBreakpointsForCurrentSource();
    }

    static string NormalizePath(string p) => Path.GetFullPath(p);

    bool VerifyLine(string path, int line)
    {
        // If no source is loaded yet, we cannot verify; treat as unverified.
        if (sourceLines.Length == 0) return false;
        var idx = Math.Clamp(line, 0, sourceLines.Length - 1);
        return !string.IsNullOrWhiteSpace(sourceLines[idx]);
    }

    public Task ContinueAsync(bool reverse = false)
    {
        // normalize instruction pointer at line start/end for the chosen direction
        if (currentLine >= 0 && currentLine < starts.Count)
        {
            instruction = reverse
                ? Math.Max(0, (ends.ElementAtOrDefault(currentLine) > 0 ? ends[currentLine] - 1 : 0))
                : starts[currentLine];
        }
        paused = false;
        while (true)
        {
            // If ExecuteLine triggers a stop event (exception/instruction/data), just return.
            if (ExecuteLine(currentLine, reverse))
            {
                return Task.CompletedTask;
            }
            // If UpdateCurrentLine signals end/start, emit end and return.
            if (UpdateCurrentLine(reverse))
            {
                debugger?.OnEnd();
                return Task.CompletedTask;
            }
            // Move to next non-empty or breakpoint line.
            if (FindNextStatement(reverse, out var reason))
            {
                if (reason == "breakpoint")
                {
                    debugger?.OnStopOnBreakpoint(currentLine, currentColumn);
                }
                return Task.CompletedTask;
            }
        }
    }



    bool ExecuteLine(int line, bool reverse = false)
    {
        // process instruction breakpoints first
        var endForLine = (line >= 0 && line < ends.Count) ? ends[line] : 0;
        var startForLine = (line >= 0 && line < starts.Count) ? starts[line] : 0;
        if (reverse)
        {
            while (instruction >= startForLine)
            {
                instruction--; // pre-decrement like TS runtime
                if (instructionBreakpoints.Contains(instruction))
                {
                    debugger?.OnStopOnInstructionBreakpoint(line, currentColumn);
                    return true;
                }
            }
        }
        else
        {

            while (instruction < endForLine)
            {
                instruction++;
                if (instructionBreakpoints.Contains(instruction))
                {
                    
                    // Ensure an output event for the breakpoint line so the adapter can position correctly
                    debugger?.OnStopOnInstructionBreakpoint(line, currentColumn);
                    return true;
                }
            }
        }

        var text = GetLine(line).Trim();
        // variable read/write detection: $name or $name=...
        foreach (Match m in VariableReadWriteRegex().Matches(text))
        {
            if (m.Groups.Count >= 2)
            {
                string? access = null;
                var name = m.Groups[1].Value;
                var hasAssign = m.Groups.Count >= 3 && m.Groups[2].Success;

                if (hasAssign)
                {
                    if (variables.Contains(name)) access = "write";
                    variables.Add(name);

                    var valToken = m.Groups.Count >= 4 ? m.Groups[3].Value : null;
                    if (!string.IsNullOrEmpty(valToken))
                    {
                        object? parsed = null;
                        if (string.Equals(valToken, "true", StringComparison.OrdinalIgnoreCase)) parsed = true;
                        else if (string.Equals(valToken, "false", StringComparison.OrdinalIgnoreCase)) parsed = false;
                        else if (valToken.StartsWith('\"') && valToken.EndsWith('\"')) parsed = valToken.Substring(1, valToken.Length - 2);
                        else if (valToken.StartsWith('{'))
                        {
                            parsed = new List<KeyValuePair<string, object?>>
                            {
                                new("fBool", true),
                                new("fInteger", 123),
                                new("fString", "hello"),
                                new("flazyInteger", 321),
                            };
                        }
                        else if (double.TryParse(valToken, System.Globalization.NumberStyles.Float, System.Globalization.CultureInfo.InvariantCulture, out var d)) parsed = d;
                        locals[name] = parsed;
                    }
                }
                else
                {
                    if (variables.Contains(name)) access = "read";
                }

                if (access != null && breakAddresses.TryGetValue(name, out var at) && at.Contains(access))
                {
                  
                    debugger?.OnStopOnDataBreakpoint(line, currentColumn);
                    return true;
                }
            }
        }

        // output events from log-like functions
        var reg1 = LogRegex();
        foreach (Match m in reg1.Matches(text))
        {
            if (m.Groups.Count == 3)
            {
                var cat = m.Groups[1].Value;
                var payload = m.Groups[2].Value;
                debugger?.OnOutput(cat, payload, sourceFile, line, m.Index);
            }
        }


        // After emitting outputs, check for exceptions so the adapter has last output line for positioning (like TS runtime)
        // named or other exceptions
        // Only match whole word 'exception' (case-insensitive) to avoid headers like 'Exceptions'.
        var namedEx = NamedExceptionRegex().Match(text);
        var hasExceptionToken = ExceptionRegex().IsMatch(text);

        if (namedEx.Success)
        {
            var ex = namedEx.Groups[1].Value.Trim();
            if (!string.IsNullOrEmpty(namedException) && string.Equals(namedException, ex, StringComparison.Ordinal))
            {
                debugger?.OnStopOnException(line, ex, currentColumn);
                return true;
            }
            else if (otherExceptions)
            {
                debugger?.OnStopOnException(line);
                return true;
            }
        }
        else if (hasExceptionToken && otherExceptions)
        {
            
            debugger?.OnStopOnException(line, null, currentColumn);
            return true;
        }

        // Do not stop here for source breakpoints; stepping moves via UpdateCurrentLine/FindNextStatement
        return false;
    }

    bool IsBreakpoint(int line)
    {
        // exact path match
        if (bps.TryGetValue(sourceFile, out var list))
        {
            return list.Any(bp => bp.Line == line);
        }
        // fallback filename match (handles path mismatches across environments)
        var name = Path.GetFileName(sourceFile);
        foreach (var kv in bps)
        {
            if (string.Equals(Path.GetFileName(kv.Key), name, StringComparison.OrdinalIgnoreCase))
            {
                return kv.Value.Any(bp => bp.Line == line);
            }
        }
        return false;
    }

    bool UpdateCurrentLine(bool reverse)
    {
        if (reverse)
        {
            if (currentLine > 0)
            {
                currentLine--;
                currentColumn = null;
                // set instruction to end of previous line
                if (currentLine >= 0 && currentLine < ends.Count)
                {
                    instruction = ends[currentLine] - 1;
                }
                return false;
            }
            else
            {
                currentLine = 0;
                currentColumn = null;
                debugger?.OnStopOnEntry(currentLine, currentColumn);
                return true;
            }
        }
        else
        {
            if (currentLine < sourceLines.Length - 1)
            {
                currentLine++;
                currentColumn = null;
                // set instruction to start of new line
                if (currentLine >= 0 && currentLine < starts.Count)
                {
                    instruction = starts[currentLine];
                }
                return false;
            }
            currentColumn = null;
            return true; // end
        }
    }

    bool FindNextStatement(bool reverse, out string reason)
    {
        reason = "breakpoint";
        if (reverse)
        {
            for (var ln = currentLine; ln >= 0; ln--)
            {
                if (IsBreakpoint(ln))
                {
                    currentLine = ln;
                    return true;
                }
                var line = GetLine(ln);
                if (!string.IsNullOrWhiteSpace(line))
                {
                    currentLine = ln;
                    break;
                }
            }
        }
        else
        {
            for (var ln = currentLine; ln < sourceLines.Length; ln++)
            {
                if (IsBreakpoint(ln))
                {
                    currentLine = ln;
                    return true;
                }
                var line = GetLine(ln);
                if (!string.IsNullOrWhiteSpace(line))
                {
                    currentLine = ln;
                    break;
                }
            }
        }
        return false;
    }

    public (IReadOnlyList<object> Frames, int Count) BuildStack(int startFrame, int endFrame)
    {
        var line = GetLine(currentLine);
        var words = GetWords(currentLine, line);
        words.Add(new Word("BOTTOM", -1, -1));

        var frames = new List<object>();
        for (int i = startFrame; i < Math.Min(endFrame, words.Count); i++)
        {
            frames.Add(new
            {
                id = i,
                name = $"{words[i].Name}({i})",
                source = new { name = Path.GetFileName(sourceFile), path = sourceFile },
                line = currentLine,
                column = currentColumn ?? 0
            });
        }
        return (frames, words.Count);
    }

    string GetLine(int? line = null)
    {
        var idx = Math.Clamp(line ?? currentLine, 0, sourceLines.Length - 1);
        return sourceLines[idx];
    }

    public string GetLineSafe(int line)
    {
        if (line < 0 || line >= sourceLines.Length) return string.Empty;
        return sourceLines[line];
    }

    static readonly Regex wordRegex = new("[a-z]+", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    static List<Word> GetWords(int l, string line)
    {
        var list = new List<Word>();
        foreach (Match m in wordRegex.Matches(line))
        {
            list.Add(new Word(m.Value, l, m.Index));
        }
        return list;
    }

    sealed record Word(string Name, int Line, int Index);

    public List<(int id, int line, bool verified)> SetBreakpoints(string path, IEnumerable<int> lines)
    {
        path = NormalizePath(path);
        var list = new List<Breakpoint>();
        bps[path] = list;
        var results = new List<(int id, int line, bool verified)>();
        foreach (var l in lines)
        {
            var verified = VerifyLine(path, l);
            var bp = new Breakpoint(nextBpId++, l, verified);
            list.Add(bp);
            results.Add((bp.Id, l, verified));
            debugger?.OnBreakpointValidated(bp .Id , l, verified);
        }
        return results;
    }

    void ReverifyBreakpointsForCurrentSource()
    {
        if (string.IsNullOrEmpty(sourceFile) || sourceLines.Length == 0) return;
        var currentName = Path.GetFileName(sourceFile);
        foreach (var kv in bps.ToArray())
        {
            var key = kv.Key;
            if (string.Equals(key, sourceFile, StringComparison.OrdinalIgnoreCase) ||
                string.Equals(Path.GetFileName(key), currentName, StringComparison.OrdinalIgnoreCase))
            {
                var list = kv.Value;
                for (int i = 0; i < list.Count; i++)
                {
                    var bp = list[i];
                    var verified = VerifyLine(sourceFile, bp.Line);
                    if (bp.Verified != verified)
                    {
                        list[i] = bp with { Verified = verified };;
                        debugger?.OnBreakpointValidated( bp.Id, bp.Line, verified );
                    }
                }
                // ensure exact-key entry points to this list
                bps.TryAdd(sourceFile, list);
            }
        }
    }

    public void Next(bool reverse = false)
    {
        // normalize instruction pointer at line start/end for the chosen direction
        if (currentLine >= 0 && currentLine < starts.Count)
        {
            instruction = reverse
                ? Math.Max(0, (ends.ElementAtOrDefault(currentLine) > 0 ? ends[currentLine] - 1 : 0))
                : starts[currentLine];
        }
        if (!ExecuteLine(currentLine, reverse))
        {
            if (!UpdateCurrentLine(reverse))
            {
                FindNextStatement(reverse, out _);
            }
            // only send step event if we didn't stop due to exception/instruction/data breakpoint
            debugger?.OnStopOnStep(currentLine, currentColumn);
        }
    }

    // Local variables APIs
    public IEnumerable<object> GetLocalVariables()
    {
        foreach (var kv in locals)
        {
            yield return new { name = kv.Key, value = ToJsonValue(kv.Value) };
        }
    }

    public void Pause()
    {
        if (paused) return;
        paused = true;
        // immediate pause: echo line and emit stopped
        var echo = GetLine(currentLine).Trim();
        debugger?.OnOutput("stdout", echo, sourceFile, currentLine, 0);
        debugger?.OnStopOnPause(currentLine, currentColumn);
    }

    public object? GetLocalVariable(string name)
    {
        if (locals.TryGetValue(name, out var v))
        {
            
            return new { name, value = ToJsonValue(v) };
        }
        return null;
    }

    public void SetVariable(string name, object? value)
    {
        locals[name] = NormalizeInboundValue(value);
    }

    static object? ToJsonValue(object? v)
    {
        if (v is List<KeyValuePair<string, object?>> list)
        {
            return list.Select(p => new { name = p.Key, value = ToJsonValue(p.Value) }).ToArray();
        }
        return v;
    }

    static object? NormalizeInboundValue(object? v)
    {
        if (v is JsonElement je)
        {
            switch (je.ValueKind)
            {
                case JsonValueKind.Null: return null;
                case JsonValueKind.True: return true;
                case JsonValueKind.False: return false;
                case JsonValueKind.Number:
                    if (je.TryGetInt64(out var l)) return (double)l;
                    if (je.TryGetDouble(out var d)) return d;
                    return null;
                case JsonValueKind.String: return je.GetString();
                case JsonValueKind.Array:
                {
                    var list = new List<KeyValuePair<string, object?>>();
                    foreach (var item in je.EnumerateArray())
                    {
                        if (item.ValueKind == JsonValueKind.Object)
                        {
                            string? name = null;
                            object? val = null;
                            foreach (var prop in item.EnumerateObject())
                            {
                                if (prop.NameEquals("name")) name = prop.Value.GetString();
                                else if (prop.NameEquals("value")) val = NormalizeInboundValue(prop.Value);
                            }
                            if (name != null)
                            {
                                list.Add(new KeyValuePair<string, object?>(name, val));
                            }
                        }
                    }
                    return list;
                }
                default:
                    return null;
            }
        }
        return v;
    }

    public IEnumerable<object> GetGlobalVariables()
    {
        yield break;
    }

    // Step in: if target provided, set column to that; else move right by 1
    public void StepIn(int? targetId)
    {
        if (targetId.HasValue)
        {
            currentColumn = targetId.Value;
        }
        else
        {
            this.currentColumn = this.currentColumn is {} currentColumn 
                ? Math.Min((sourceLines[currentLine].Length), currentColumn + 1) 
                : 1;
        }
        debugger?.OnStopOnStep(currentLine);
    }

    // Step out: move left by 1; if reaches 0 unset column
    public void StepOut()
    {
        if (currentColumn.HasValue)
        {
            var next = currentColumn.Value - 1;
            currentColumn = next <= 0 ? null : next;
        }
        debugger?.OnStopOnStep(currentLine);
    }

    // Column breakpoint positions for a line: start index of words with name length > 8 (like TS)
    public List<int> GetBreakpointColumns(string path, int line)
    {
        // If no source loaded, return empty to avoid index errors
        if (sourceLines.Length == 0) return new List<int>();
        var words = GetWords(line, GetLine(line));
        return words.Where(w => w.Name.Length > 8).Select(w => w.Index).ToList();
    }

    // Disassemble: return list of words from global instruction stream
    public IEnumerable<object> Disassemble(int address, int instructionCount)
    {
        var list = new List<object>();
        for (int a = address; a < address + instructionCount; a++)
        {
            if (a >= 0 && a < instructions.Count)
            {
                var w = instructions[a];
                list.Add(new { address = a, instruction = w.Name, line = w.Line });
            }
            else
            {
                list.Add(new { address = a, instruction = "nop" });
            }
        }
        return list;
    }

    // Data breakpoints
    public bool SetDataBreakpoint(string address, string accessType)
    {
        if (string.IsNullOrWhiteSpace(address)) return false;
        var norm = accessType == "readWrite" ? "read write" : accessType;
        if (breakAddresses.TryGetValue(address, out var existing))
        {
            if (!existing.Contains(norm)) breakAddresses[address] = "read write";
        }
        else
        {
            breakAddresses[address] = norm;
        }
        return true;
    }

    public void ClearAllDataBreakpoints() => breakAddresses.Clear();

    // Instruction breakpoints
    public bool SetInstructionBreakpoint(int address) => instructionBreakpoints.Add(address);
    public void ClearInstructionBreakpoints() => instructionBreakpoints.Clear();

    // Exception filters
    public void SetExceptionsFilters(string? namedException, bool otherExceptions)
    {
        this.namedException = namedException;
        this.otherExceptions = otherExceptions;
    }

    [GeneratedRegex(@"\$([a-z][a-z0-9]*)(=(false|true|[0-9]+(\.[0-9]+)?|""[^""]*""|\{.*\}))?", RegexOptions.IgnoreCase, "ja-JP")]
    private static partial Regex VariableReadWriteRegex();
    [GeneratedRegex(@"(log|prio|out|err)\(([^)]*)\)")]
    private static partial Regex LogRegex();
    [GeneratedRegex("exception\\((.*)\\)", RegexOptions.IgnoreCase, "ja-JP")]
    private static partial Regex NamedExceptionRegex();
    [GeneratedRegex(@"\bexception\b", RegexOptions.IgnoreCase, "ja-JP")]
    private static partial Regex ExceptionRegex();
}

File: mock-go/internal/engine/engine.go
package engine

import (
    "bufio"
    "path/filepath"
    "regexp"
    "strings"
)

type Debugger interface {
    OnStopOnEntry(line int, column *int)
    OnStopOnStep(line int, column *int)
    OnStopOnBreakpoint(line int, column *int)
    OnStopOnException(line int, exception *string, column *int)
    OnStopOnDataBreakpoint(line int, column *int)
    OnStopOnInstructionBreakpoint(line int, column *int)
    OnStopOnPause(line int, column *int)
    OnBreakpointValidated(id int, verified bool)
    OnOutput(category, text, file string, line, column int)
    OnEnd()
}

type Breakpoint struct {
    ID       int
    Line     int
    Verified bool
}

type Word struct {
    Name  string
    Line  int
    Index int
}

type Engine struct {
    dbg Debugger

    sourceFile  string
    sourceLines []string

    currentLine  int
    currentCol   *int
    instruction  int
    instructions []Word
    starts       []int
    ends         []int

    nextBpID int
    bps      map[string][]Breakpoint

    namedException  *string
    otherExceptions bool

    dataBps   map[string]string
    instrBps  map[int]struct{}

    variables map[string]struct{}
    locals    map[string]any

    paused bool
}

func New(d Debugger) *Engine {
    return &Engine{
        dbg:        d,
        currentLine: 0,
        bps:        map[string][]Breakpoint{},
        dataBps:    map[string]string{},
        instrBps:   map[int]struct{}{},
        variables:  map[string]struct{}{},
        locals:     map[string]any{},
        nextBpID:   1,
    }
}

func (e *Engine) SourceFile() string { return e.sourceFile }
func (e *Engine) SourceLength() int { return len(e.sourceLines) }

func (e *Engine) LoadSource(path string, contents []byte) {
    e.sourceFile = abs(path)
    e.sourceLines = splitLines(string(contents))
    e.currentLine = 0
    e.currentCol = nil
    e.instructions = e.instructions[:0]
    e.starts = e.starts[:0]
    e.ends = e.ends[:0]
    for i, line := range e.sourceLines {
        _ = i
        e.starts = append(e.starts, len(e.instructions))
        words := getWords(i, line)
        e.instructions = append(e.instructions, words...)
        e.ends = append(e.ends, len(e.instructions))
    }
    if len(e.starts) > 0 {
        e.instruction = e.starts[0]
    } else { e.instruction = 0 }
}

func (e *Engine) Pause() {
    if e.paused {
        return
    }
    e.paused = true
    // if not running a loop, emit immediately
    e.dbg.OnOutput("stdout", strings.TrimSpace(e.getLine(e.currentLine)), e.sourceFile, e.currentLine, 0)
    e.dbg.OnStopOnPause(e.currentLine, e.currentCol)
}

func (e *Engine) Continue(reverse bool) {
    // normalize instruction at start/end of current line
    if e.currentLine >= 0 && e.currentLine < len(e.starts) {
        if reverse {
            end := e.ends[e.currentLine]
            if end > 0 { e.instruction = end-1 } else { e.instruction = 0 }
        } else {
            e.instruction = e.starts[e.currentLine]
        }
    }
    e.paused = false
    for {
        if e.paused {
            e.paused = false
            e.dbg.OnStopOnPause(e.currentLine, e.currentCol)
            return
        }
        if e.executeLine(e.currentLine, reverse) {
            return
        }
        if e.updateCurrentLine(reverse) {
            e.dbg.OnEnd()
            return
        }
        if e.findNextStatement(reverse) {
            return
        }
    }
}

func (e *Engine) Next(reverse bool) {
    if e.currentLine >= 0 && e.currentLine < len(e.starts) {
        if reverse {
            end := e.ends[e.currentLine]
            if end > 0 { e.instruction = end-1 } else { e.instruction = 0 }
        } else {
            e.instruction = e.starts[e.currentLine]
        }
    }
    if !e.executeLine(e.currentLine, reverse) {
        if !e.updateCurrentLine(reverse) {
            e.findNextStatement(reverse)
        }
        e.dbg.OnStopOnStep(e.currentLine, e.currentCol)
    }
}

func (e *Engine) StepIn(targetID *int) {
    if targetID != nil {
        e.currentCol = targetID
    } else {
        if e.currentCol != nil {
            v := *e.currentCol + 1
            e.currentCol = &v
        } else {
            v := 1
            e.currentCol = &v
        }
    }
    e.dbg.OnStopOnStep(e.currentLine, e.currentCol)
}

func (e *Engine) StepOut() {
    if e.currentCol != nil {
        v := *e.currentCol - 1
        if v <= 0 { e.currentCol = nil } else { e.currentCol = &v }
    }
    e.dbg.OnStopOnStep(e.currentLine, e.currentCol)
}

func (e *Engine) BuildStack(start, end int) (frames []map[string]any, count int) {
    line := e.getLine(e.currentLine)
    words := getWords(e.currentLine, line)
    words = append(words, Word{Name: "BOTTOM", Line: -1, Index: -1})
    column := 0
    if e.currentCol != nil { column = *e.currentCol }
    for i := start; i < min(end, len(words)); i++ {
        frames = append(frames, map[string]any{
            "id":    i,
            "name":  words[i].Name + "(" + itoa(i) + ")",
            "source": map[string]any{"name": basename(e.sourceFile), "path": e.sourceFile},
            "line":  e.currentLine,
            "column": column,
        })
    }
    return frames, len(words)
}

func (e *Engine) SetBreakpoints(path string, lines []int) (res []map[string]any) {
    p := abs(path)
    list := make([]Breakpoint, 0, len(lines))
    e.bps[p] = list
    for _, l := range lines {
        verified := e.verifyLine(p, l)
        bp := Breakpoint{ID: e.nextBpID, Line: l, Verified: verified}
        e.nextBpID++
        e.bps[p] = append(e.bps[p], bp)
        e.dbg.OnBreakpointValidated(bp.ID, verified)
        res = append(res, map[string]any{"id": bp.ID, "verified": verified, "line": l})
    }
    return
}

func (e *Engine) GetBreakpointColumns(_path string, line int) []int {
    cols := []int{}
    for _, w := range getWords(line, e.getLine(line)) {
        if len(w.Name) > 8 { cols = append(cols, w.Index) }
    }
    return cols
}

func (e *Engine) GetBreakpointLines() []int {
    var out []int
    for i := range e.sourceLines {
        if strings.TrimSpace(e.sourceLines[i]) != "" { out = append(out, i) }
    }
    return out
}

func (e *Engine) Disassemble(address, count int) []map[string]any {
    var list []map[string]any
    for a := address; a < address+count; a++ {
        if a >= 0 && a < len(e.instructions) {
            w := e.instructions[a]
            list = append(list, map[string]any{"address": a, "instruction": w.Name, "line": w.Line})
        } else {
            list = append(list, map[string]any{"address": a, "instruction": "nop"})
        }
    }
    return list
}

// Variables & breakpoints APIs
func (e *Engine) GetLocalVariables() []map[string]any {
    out := []map[string]any{}
    for k, v := range e.locals {
        out = append(out, map[string]any{"name": k, "value": v})
    }
    return out
}

func (e *Engine) GetLocalVariable(name string) map[string]any {
    if v, ok := e.locals[name]; ok { return map[string]any{"name": name, "value": v} }
    return nil
}

func (e *Engine) SetVariable(name string, value any) { e.locals[name] = value }

func (e *Engine) GetGlobalVariables() []map[string]any {
    out := []map[string]any{}
    for i := 0; i < 10; i++ { out = append(out, map[string]any{"name": "global_" + itoa(i), "value": i}) }
    return out
}

func (e *Engine) SetExceptionsFilters(named *string, others bool) {
    e.namedException = named
    e.otherExceptions = others
}

func (e *Engine) SetDataBreakpoint(address, access string) bool {
    if access == "readWrite" { access = "read write" }
    if cur, ok := e.dataBps[address]; ok {
        if cur != access { e.dataBps[address] = "read write" }
    } else { e.dataBps[address] = access }
    return true
}
func (e *Engine) ClearAllDataBreakpoints() { e.dataBps = map[string]string{} }

func (e *Engine) SetInstructionBreakpoint(addr int) bool { e.instrBps[addr] = struct{}{}; return true }
func (e *Engine) ClearInstructionBreakpoints() { e.instrBps = map[int]struct{}{} }

// helpers
func (e *Engine) verifyLine(_path string, line int) bool {
    if line < 0 || line >= len(e.sourceLines) { return false }
    return strings.TrimSpace(e.sourceLines[line]) != ""
}

func (e *Engine) getLine(line int) string {
    if line < 0 || line >= len(e.sourceLines) { return "" }
    return e.sourceLines[line]
}

func (e *Engine) updateCurrentLine(reverse bool) bool {
    if reverse {
        if e.currentLine > 0 { e.currentLine-- } else { e.currentLine = 0; e.currentCol = nil; e.dbg.OnStopOnEntry(e.currentLine, e.currentCol); return true }
    } else {
        if e.currentLine < len(e.sourceLines)-1 { e.currentLine++ } else { e.currentCol = nil; return true }
    }
    return false
}

func (e *Engine) findNextStatement(reverse bool) bool {
    for ln := e.currentLine; ; {
        // line bp
        if list, ok := e.bps[e.sourceFile]; ok {
            for _, bp := range list {
                if bp.Line == ln {
                    if !bp.Verified { bp.Verified = true; e.dbg.OnBreakpointValidated(bp.ID, true) }
                    e.currentLine = ln
                    e.dbg.OnStopOnBreakpoint(e.currentLine, e.currentCol)
                    return true
                }
            }
        }
        // instr bp at line start/end
        addr := 0
        if reverse { addr = e.starts[ln] } else { addr = e.ends[ln]-1 }
        if _, ok := e.instrBps[addr]; ok { e.currentLine = ln; e.dbg.OnStopOnInstructionBreakpoint(e.currentLine, e.currentCol); return true }

        line := strings.TrimSpace(e.getLine(ln))
        if line != "" { e.currentLine = ln; break }
        if reverse { if ln <= 0 { break }; ln-- } else { ln++; if ln >= len(e.sourceLines) { break } }
    }
    return false
}

var (
    wordRe   = regexp.MustCompile(`[a-zA-Z]+`)
    rwVarRe  = regexp.MustCompile(`\$([a-zA-Z][a-zA-Z0-9]*)(=(false|true|[0-9]+(\.[0-9]+)?|\".*\"|\{.*\}))?`)
    logRe    = regexp.MustCompile(`(log|prio|out|err)\(([^\)]*)\)`)
    excName  = regexp.MustCompile(`exception\((.*)\)`)
    excToken = regexp.MustCompile(`\bexception\b`)
)

func (e *Engine) executeLine(ln int, reverse bool) bool {
    // instruction breakpoints first
    start := e.starts[ln]
    end := e.ends[ln]
    if reverse {
        for e.instruction >= start {
            e.instruction--
            if _, ok := e.instrBps[e.instruction]; ok { e.dbg.OnStopOnInstructionBreakpoint(ln, e.currentCol); return true }
        }
    } else {
        for e.instruction < end {
            e.instruction++
            if _, ok := e.instrBps[e.instruction]; ok { e.dbg.OnStopOnInstructionBreakpoint(ln, e.currentCol); return true }
        }
    }

    text := strings.TrimSpace(e.getLine(ln))

    // variable read/write; data breakpoints
    ms := rwVarRe.FindAllStringSubmatchIndex(text, -1)
    for _, idx := range ms {
        name := text[idx[2]:idx[3]]
        hasAssign := idx[4] >= 0
        var access *string
        if hasAssign {
            if _, ok := e.variables[name]; ok { s := "write"; access = &s }
            e.variables[name] = struct{}{}
            // capture value token if present; set locals loosely
            if idx[6] >= 0 {
                token := text[idx[6]:idx[7]]
                e.locals[name] = parseToken(token)
            }
        } else {
            if _, ok := e.variables[name]; ok { s := "read"; access = &s }
        }
        if access != nil {
            if flg, ok := e.dataBps[name]; ok && strings.Contains(flg, *access) {
                e.dbg.OnStopOnDataBreakpoint(ln, e.currentCol)
                return true
            }
        }
    }

    // outputs
    for _, m := range logRe.FindAllStringSubmatchIndex(text, -1) {
        if len(m) >= 6 {
            cat := text[m[2]:m[3]]
            payload := text[m[4]:m[5]]
            e.dbg.OnOutput(cat, payload, e.sourceFile, ln, m[0])
        }
    }

    // exceptions
    if m := excName.FindStringSubmatch(text); len(m) == 2 {
        ex := strings.TrimSpace(m[1])
        if e.namedException != nil && *e.namedException == ex { e.dbg.OnStopOnException(ln, &ex, e.currentCol); return true }
        if e.otherExceptions { e.dbg.OnStopOnException(ln, nil, e.currentCol); return true }
    } else {
        if excToken.MatchString(text) && e.otherExceptions { e.dbg.OnStopOnException(ln, nil, e.currentCol); return true }
    }

    return false
}

// utils
func splitLines(s string) []string {
    sc := bufio.NewScanner(strings.NewReader(s))
    lines := []string{}
    for sc.Scan() { lines = append(lines, sc.Text()) }
    // append last empty line? not needed for our semantics
    return lines
}

func getWords(l int, line string) []Word {
    out := []Word{}
    for _, m := range wordRe.FindAllStringSubmatchIndex(line, -1) {
        out = append(out, Word{Name: line[m[0]:m[1]], Line: l, Index: m[0]})
    }
    return out
}

func parseToken(t string) any {
    switch {
    case t == "true":
        return true
    case t == "false":
        return false
    case strings.HasPrefix(t, "\"") && strings.HasSuffix(t, "\""):
        return strings.Trim(t, "\"")
    case strings.HasPrefix(t, "{"):
        return []map[string]any{{"name": "fBool", "value": true}, {"name": "fInteger", "value": 123}, {"name": "fString", "value": "hello"}, {"name": "flazyInteger", "value": 321}}
    default:
        // try int/float
        // keep as string if parse not needed; adapter treats primitives loosely
        return t
    }
}

func abs(p string) string { a, _ := filepath.Abs(p); return a }
func basename(p string) string { return filepath.Base(p) }
func min(a, b int) int { if a < b { return a } ; return b }
func itoa(i int) string { return strconvItoa(i) }

// tiny itoa without importing strconv everywhere
func strconvItoa(i int) string {
    // simple base-10 conversion
    if i == 0 { return "0" }
    neg := false
    if i < 0 { neg = true; i = -i }
    var buf [20]byte
    bp := len(buf)
    for i > 0 { bp--; buf[bp] = byte('0' + (i % 10)); i /= 10 }
    if neg { bp--; buf[bp] = '-' }
    return string(buf[bp:])
}
